nexus/db-queries/src/db/queries/affinity.rs

```diff
@@ -0,0 +1,608 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Implementation of queries for affinity groups
+
+use crate::db::model::AffinityPolicyEnum;
+use crate::db::raw_query_builder::{QueryBuilder, TypedSqlQuery};
+use diesel::sql_types;
+use omicron_uuid_kinds::GenericUuid;
+use omicron_uuid_kinds::InstanceUuid;
+
+/// For an instance, look up all anti-affinity groups, and return
+/// a list of sleds with other instances in that anti-affinity group
+/// already reserved.
+pub fn lookup_anti_affinity_sleds_query(
+    instance_id: InstanceUuid,
+) -> TypedSqlQuery<(AffinityPolicyEnum, sql_types::Uuid)> {
+    QueryBuilder::new().sql(
+        "WITH our_groups AS (
+            SELECT group_id
+            FROM anti_affinity_group_instance_membership
+            WHERE instance_id = ").param().sql("
+         ),
+         other_instances AS (
+            SELECT anti_affinity_group_instance_membership.group_id,instance_id
+            FROM anti_affinity_group_instance_membership
+            JOIN our_groups
+            ON anti_affinity_group_instance_membership.group_id = our_groups.group_id
+            WHERE instance_id != ").param().sql("
+         ),
+         other_instances_by_policy AS (
+            SELECT policy,instance_id
+            FROM other_instances
+            JOIN anti_affinity_group
+            ON
+                anti_affinity_group.id = other_instances.group_id AND
+                anti_affinity_group.failure_domain = 'sled'
+            WHERE anti_affinity_group.time_deleted IS NULL
+         )
+         SELECT DISTINCT policy,sled_id
+         FROM other_instances_by_policy
+         JOIN sled_resource
+         ON
+            sled_resource.instance_id = other_instances_by_policy.instance_id AND
+            sled_resource.kind = 'instance'")
+     .bind::<sql_types::Uuid, _>(instance_id.into_untyped_uuid())
+     .bind::<sql_types::Uuid, _>(instance_id.into_untyped_uuid())
+     .query()
+}
```

Comment:
didn't know we were allowed to do this 😈

Reply:
This style of query builder? I added this in https://github.com/oxidecomputer/omicron/pull/5063 , and ripped out the old one in https://github.com/oxidecomputer/omicron/pull/5966

CTEs seem tricky enough without fighting the type system to make them happen

Reply:
AFAIK `diesel` just straight up doesn't _have_ a way to express CTEs?